### PR TITLE
typed Parameterized class

### PR DESCRIPTION
**Pull request by tscholak**
- [x] implement `Torch.Typed.Parameter.Parameterized` with two member functions: `flattenParameters` and `replaceParameters`. The former returns a fully typed `HList` instead of a list, the latter takes a model `f` and an `HList` and returns an updated `f` while avoiding the `State` monad used in `Torch.NN`.
- [x] remove old `Torch.NN.Parameterized` instances from the `Torch.Typed` modules.
- [x] adapt the `Torch.Typed.NN.Recurrent.LSTM` module to the new typed `Parameterized` class. This is partially based on unreviewed work by @o1lo01ol1o, see https://github.com/hasktorch/hasktorch/pull/243.
- [ ] adapt examples

Here's a script you can run in ghci:
```
:set prompt "λ> "
:seti -XTypeApplications
:seti -XDataKinds
:seti -XScopedTypeVariables
import Data.HList
import Torch.Typed.NN.Recurrent.LSTM
import Torch.Typed.Tensor
import Torch.Typed.Parameter
import Torch.Typed.NN
import Torch.Typed.Native
import Torch.Typed.Factories
import qualified Torch.DType as D
import qualified Torch.Device as D
import qualified Torch.NN as A
spec = LSTMSpec @5 @7 @3 @'Bidirectional @'D.Float @'( 'D.CPU, 0) (DropoutSpec 0.1)
model <- A.sample spec
:t flattenParameters model
spec' = LSTMWithConstInitSpec spec zeros zeros
model' <- A.sample spec'
:t flattenParameters model'
:t Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model' zeros
:t Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model' zeros
spec'' = LSTMWithLearnedInitSpec spec zeros zeros
model'' <- A.sample spec''
:t flattenParameters model''
:t Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model'' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model'' zeros
:t Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model'' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model'' zeros
```

and here is its output:
```
*Data.HList> :set prompt "λ> "
λ> :seti -XTypeApplications
λ> :seti -XDataKinds
λ> :seti -XScopedTypeVariables
λ> import Data.HList
λ> import Torch.Typed.NN.Recurrent.LSTM
λ> import Torch.Typed.Tensor
λ> import Torch.Typed.Parameter
λ> import Torch.Typed.NN
λ> import Torch.Typed.Native
λ> import Torch.Typed.Factories
λ> import qualified Torch.DType as D
λ> import qualified Torch.Device as D
λ> import qualified Torch.NN as A
λ> spec = LSTMSpec @5 @7 @3 @'Bidirectional @'D.Float @'( 'D.CPU, 0) (DropoutSpec 0.1)
λ> model <- A.sample spec
λ> :t flattenParameters model
flattenParameters model
  :: HList
       '[Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28]]
λ> spec' = LSTMWithConstInitSpec spec zeros zeros
λ> model' <- A.sample spec'
λ> :t flattenParameters model'
flattenParameters model'
  :: HList
       '[Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28]]
λ> :t Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model' zeros
  :: (Tensor '( 'D.CPU, 0) 'D.Float '[1, 4, 14],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7])
λ> :t Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model' zeros
  :: (Tensor '( 'D.CPU, 0) 'D.Float '[4, 1, 14],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7])
λ> spec'' = LSTMWithLearnedInitSpec spec zeros zeros
λ> model'' <- A.sample spec''
λ> :t flattenParameters model''
flattenParameters model''
  :: HList
       '[Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 5],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 14],
         Parameter '( 'D.CPU, 0) 'D.Float '[28, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[28],
         Parameter '( 'D.CPU, 0) 'D.Float '[6, 7],
         Parameter '( 'D.CPU, 0) 'D.Float '[6, 7]]
λ> :t Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model'' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model'' zeros
  :: (Tensor '( 'D.CPU, 0) 'D.Float '[1, 4, 14],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7])
λ> Torch.Typed.NN.Recurrent.LSTM.lstm @'BatchFirst @1 @4 False model'' zeros
(Tensor Float [1,4,14] ,Tensor Float [6,1,7] ,Tensor Float [6,1,7] )
λ> :t Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model'' zeros
Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model'' zeros
  :: (Tensor '( 'D.CPU, 0) 'D.Float '[4, 1, 14],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7],
      Tensor '( 'D.CPU, 0) 'D.Float '[6, 1, 7])
λ> Torch.Typed.NN.Recurrent.LSTM.lstm @'SequenceFirst @1 @4 False model'' zeros
(Tensor Float [4,1,14] ,Tensor Float [6,1,7] ,Tensor Float [6,1,7] )
λ>
```

### CODE REVIEW

File: hasktorch/src/Torch/Typed/NN.hs [hunk 1]
```diff
@@ -247,7 +244,7 @@ conv1d Conv1d {..} input = Torch.Typed.Native.conv1d @stride @padding
   (toDependent conv1dBias)
   input
 
-instance A.Parameterized (Conv1d inputChannelSize outputChannelSize kernelSize dtype device)
+-- instance A.Parameterized (Conv1d inputChannelSize outputChannelSize kernelSize dtype device)
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Typed/NN.hs]: New typed Parameterized class does not need this instance?

Comment (tscholak) [reply hasktorch/src/Torch/Typed/NN.hs]: no, it's fully automatically derived for Models with a `Generic` instance.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/NN.hs]: It is great.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/NN.hs]: Do you have any reason to use comment-out?
I think we can delete it.

Comment (tscholak) [reply hasktorch/src/Torch/Typed/NN.hs]: actually, I decided to add these instances back.
I looked at the additional work necessary to port the examples to the new typed `Parameterized` class, and I think it warrants another PR.
so let's keep these instance around a little while longer.

Comment (junjihashimoto) [reply hasktorch/src/Torch/Typed/NN.hs]: 🆗

**Pull request merged**